collectors/plugins.d/plugins_d.c

```diff
@@ -246,6 +277,129 @@ inline size_t pluginsd_process(RRDHOST *host, struct plugind *cd, FILE *fp, int
+    static RRDSET   *__stream_prs_st = NULL;     //       for active processors
+    static RRDSET   *__stream_metrics_st = NULL; //       for active metrics we collect from slaves
+    static RRDSET   *__stream_processes_st = NULL;  //    for total active processes in slaves
+    static time_t   stream_updated   = 0;           // Used so that we Do not rrdset_done in less than a second
+    time_t          thread_start_time;
+    static int      stream_stats_activated = 0;
+    
+	if (likely(stream_stats_enabled) && likely(!stream_stats_activated) && likely(host != localhost)) {
+        stream_stats_activated = 1;
+
+        __stream_mem_st = rrdset_create_localhost(
+                "netdata"
+                , "memory" 
+                , NULL
+                , "streams"
+                , NULL
+                , "NetData Incoming Streams"
+                , "MB"
+                , "netdata"
+                , "stats"
+                , 130000
+                , localhost->rrd_update_every
+                , RRDSET_TYPE_LINE
+            );
+        __stream_cnt_st = rrdset_create_localhost(
+                "netdata"
+                , "stream count" 
+                , "streams"
+                , "streams"
+                , NULL
+                , "NetData Incoming Streams"
+                , "Count"
+                , "netdata"
+                , "stats"
+                , 130001
+                , localhost->rrd_update_every
+                , RRDSET_TYPE_LINE
+            );
+            __stream_prs_st = rrdset_create_localhost(
+                "netdata"
+                , "active processors" 
+                , "streams"
+                , "streams"
+                , NULL
+                , "NetData Incoming Streams"
+                , "Count"
+                , "netdata"
+                , "stats"
+                , 130002
+                , localhost->rrd_update_every
+                , RRDSET_TYPE_LINE
+            );
+            __stream_metrics_st = rrdset_create_localhost(
+                "netdata"
+                , "metrics" 
+                , "streams"
+                , "streams"
+                , NULL
+                , "NetData Incoming Streams"
+                , "Count"
+                , "netdata"
+                , "stats"
+                , 130003
+                , localhost->rrd_update_every
+                , RRDSET_TYPE_LINE
+            );
+            __stream_processes_st = rrdset_create_localhost(
+                "netdata"
+                , "processes" 
+                , "streams"
+                , "streams"
+                , NULL
+                , "NetData Incoming Streams"
+                , "Count"
+                , "netdata"
+                , "stats"
+                , 130004
+                , localhost->rrd_update_every
+                , RRDSET_TYPE_LINE
+            );
+
+            if (likely(__stream_mem_st)) 
```

Comment:
I would use an array literal of `static struct`s here and iterate over it instead of having several repeating if statements. It’s cleaner, less prone to error, more maintainable and the performance impact is minimal.

Reply:
Hi @cosmix 

You are right. It would make more sense to have an iteration and it will be the way to go if this is to be further expanded -- it can easily have half a dozen of more charts to process and it would be a nightmare to review and maintain. 

In addiction to that it would almost certainly go beyond static structures and require additional book keeping (memory allocation) -- image just these 5 charts grouped by some sort of OS / Kernel version or CPU arch when having a couple of hundred slaves 

All in all  yes !  this would would be rewritten